### PR TITLE
Added new messages for stateful recovery

### PR DESCRIPTION
**Pull request by srkukarni**

### CODE REVIEW

**Comment by kramasamy** [comment]
@maosongfu - thanks for the write up. I like push model because 

- It is more deterministic (unlike pull - where you have to poll the store periodically to see if the checkpoint is complete and polling could put stress on Name Node for HDFS)
- It is independent of the underlying storage (this is a great point, as you mentioned)

File: heron/proto/ckptmgr.proto [hunk 1]
```diff
@@ -5,11 +5,33 @@ option java_outer_classname = "CheckpointManager";
 import "common.proto";
 import "physical_plan.proto";
 
+// message stored in zk by the tmaster
+message StatefulMostRecentCheckpoint {
```
Comment (srkukarni) [inline heron/proto/ckptmgr.proto]: We need to keep things backwards compatible as much as possible. Thus I'm not in the favor of altering TMasterLocation data structure. My inclination is to go for another top level directory

File: heron/proto/ckptmgr.proto [hunk 2]
```diff
@@ -5,11 +5,33 @@ option java_outer_classname = "CheckpointManager";
...
+// to this checkpoint
+message ResetTopologyState {
+  required string checkpoint_id = 1;
+}
+
+// Message sent by stmgr to tmaster asking it to reset the topology
+// to some valid checkpoint. This is sent either if stmgr dies
+// and comes back up or if an instance dies.
+message ResetTopologyStateRequest {
+  required string reason = 1;
```
Comment (srkukarni) [inline heron/proto/ckptmgr.proto]: We could

**Comment by kramasamy** [comment]
For reason both int and string will be nice. If you have int only then we have to maintain the mapping into to string

**Comment by srkukarni** [comment]
Please check now

File: heron/proto/ckptmgr.proto [hunk 3]
```diff
@@ -5,11 +5,41 @@ option java_outer_classname = "CheckpointManager";
...
+// Message sent by tmaster to stmgr asking them to reset their instances
+// to this checkpoint
+message RestoreTopologyState {
+  required string checkpoint_id = 1;
+}
+
+// Message sent by stmgr to tmaster asking it to reset the topology
+// to some valid checkpoint. This is sent either if stmgr dies
+// and comes back up or if an instance dies.
+message RestoreTopologyStateRequest {
```
Comment (maosongfu) [inline heron/proto/ckptmgr.proto]: Can this request simply be kept with best effort semantics? Need we add a response for acknowledgement to support at least once semantics for this request?

Failure to deliver message to tmaster will be detected by later heartbeat. It seems ok to keep this request with only best effort semantics though the handling and reasoning can be tricky.
Any thoughts?

File: heron/proto/ckptmgr.proto [hunk 4]
```diff
@@ -47,13 +64,51 @@ message SaveStateCheckpoint {
...
+// This is the message that the instance sends
+// when it wants stmgr to store its state
+message InstanceStateCheckpoint {
+  required string checkpoint_id = 1;
+  required bytes state = 2;
+}
+
+// This is the message that stmgr sends to its instace
+// asking them to restore their state
+message RestoreInstanceState {
```
Comment (maosongfu) [inline heron/proto/ckptmgr.proto]: Could this request simply be kept with best effort semantics? Need we add a response for acknowledgement to support at least once semantics for this request?

This one is slightly different from the request sending from the request sent from stmgr to tmaster: stmgr and instances are on the same host and are bind with long live connections. Potential failure to lose this message can be detected and leads to another re-wind.
It seems ok to keep this request with only best effort semantics though the handling and reasoning can be tricky.
Any thoughts?

File: heron/proto/ckptmgr.proto (hunk 3 above)
Comment (srkukarni) [inline heron/proto/ckptmgr.proto]: Its not a request, but I agree that the name is confusing. Let me change this

File: heron/proto/ckptmgr.proto [hunk 5]
```diff
@@ -5,11 +5,41 @@ option java_outer_classname = "CheckpointManager";
...
+// to some valid checkpoint. This is sent either if stmgr dies
+// and comes back up or if an instance dies.
+message RestoreTopologyStateRequest {
+  required string reason = 1;
+}
+
+// Message sent by stmgr to tmaster informing it about
+// the fact that we stored a checkpoint belonging
+// to the instance
+message TopologyStateStored {
```
Comment (srkukarni) [inline heron/proto/ckptmgr.proto]: I don't like this. This introduces unnecessary logic inside stmgr. How about stmgr just passing this message to the tmaster with just the instance task id and the tmaster determining things

File: heron/proto/ckptmgr.proto (hunk 3 above)
Comment (maosongfu) [inline heron/proto/ckptmgr.proto]: Then what happen if this message fails to be delivered to tmaster?

File: heron/proto/ckptmgr.proto (hunk 5 above)
Comment (maosongfu) [inline heron/proto/ckptmgr.proto]: sounds good.

File: heron/proto/ckptmgr.proto (hunk 4 above)
Comment (maosongfu) [inline heron/proto/ckptmgr.proto]: If sending with only best effot, then what happen if this message fails to be delivered to instance?

**Comment by maosongfu** [comment]
In fact, the actual concern is:
What should we do if we fail to deliver RestoreTopologyState, RestoreTopologyStateMessage  and RestoreInstanceState to their target? We need a proper failure recovery mechanism if trying to send them with simply best effort.

**Comment by maosongfu** [comment]
I re-reasoned it carefully. And it can work to send these 3 messages with only best effort if we rewind the whole topology in any failure happens, including tmaster, stmgr, instance failures.

**Comment by maosongfu** [comment]
looks good to me. 👍 
(Except the names of message are complicated, not that informative and not easy to distinguish/refer...)

**Pull request merged**